clang/lib/Sema/SemaStmt.cpp

```diff
@@ -3371,6 +3379,20 @@ Sema::ActOnBreakStmt(SourceLocation BreakLoc, Scope *CurScope) {
   if (S->isOpenMPLoopScope())
     return StmtError(Diag(BreakLoc, diag::err_omp_loop_cannot_use_stmt)
                      << "break");
+
+  // OpenACC doesn't allow 'break'ing from a compute construct, so diagnose if
+  // we are trying to do so.  This can come in 2 flavors: 1-the break'able thing
+  // (besides the compute construct) 'contains' the compute construct, at which
+  // point the 'break' scope will be the compute construct.  Else it could be a
+  // loop of some sort that has a direct parent of the compute construct.
+  // However, a 'break' in a 'switch' marked as a compute construct doesn't
+  // count as 'branch out of' the compute construct.
+  if (S->isOpenACCComputeConstructScope() ||
+      (!S->isDirectlySwitchScope() && S->getParent() &&
```

Comment:
Why do you need to check for non-switch scope here? Is it by default, that if it is openacc block, it is not switch scope? Because you create openacc scope explicitly without setting switch scope, no?

Reply:
This is covering 2 separate cases.

First:


```
switch(F) { <-- Switch scope
case 1:
#pragma acc parallel <-- compute-construct scope
break;
}
```
In this case, this is the pre-'||' check, the pragma's scope itself (created in ParseOpenACC.cpp) will be the 'break parent' of current, which cannot be broken from (so it is diagnosed).  This can NEVER be a 'switch' scope, as it is a 'pragma acc compute-construct' scope.

Second:

```
#pragma acc parallel <-- compute construct scope
switch(F) { <-- switch scope
case 1:
break;
}
```
This is how the OpenMP variant of this diagnostic works, the 'scope' here is the 'switch scope'.  Additionally, the 'parent' of this one is the Compute Construct scope.  So everything after the '||' handles this. (note that `isOpenMPLoopScope` handles this by implicitly checking the 'parent').  Since we have the 'switch' scope (or other loop/scope/etc), we have to check that it is directly inside of the pragma scope.  Since this one CAN be a 'switch'/any other loop/scope/etc, we have to check it, as well as checking the 'parent' for being an OpenACC Compute construct.

Reply:
You just need to check that this is a loop scope with the parent openacc scope. break should not be allowed in loops only, right? Why do you check for not-switches? Can you check instead for the loop scope explicitly? Much easier to read and maintain.

Reply:
OpenACC compute constructs are NOT limited to loops unfortunately, they can be any statement, so just checking the loop isn't sufficient.  And checking the 'is loop with a parent openacc scope' doesn't work if the openacc-pragma is inside the loop.  See my 1st example above, the pragma is the 'immediate' scope, but the 'switch' is above it (or the loop in that case).

Reply:
The first (dangling) case should be handled by getBreakParent() function. OpenMP set FnScope (IIRC) to nullify break parent and diagnose dangling break/continue and other similar constructs.

Reply:
> The first (dangling) case should be handled by getBreakParent() function. OpenMP set FnScope (IIRC) to nullify break parent and diagnose dangling break/continue and other similar constructs.

I saw that, but it ends up being much worse diagnostics, doesn't it?  A case like above just says "error: 'break' statement not in loop or switch statement", which seems inaccurate for OpenACC given the 'ignorability' requirement of the language.

Reply:
ALSO, it doesn't help the 'continue' case, where the inverse is the requirement (a continue on the 'annotated loop' is perfectly fine, but not one that would jump out of the pragma).

Reply:
>What is "ignorability"? Sorry, just asking to understand the context more correctly.

The idea is that we should be able to 'ignore' the pragmas and have the program still have meaning.  A big part of the 'theory' here is that you should be able to add pragmas to existing code and get 'descriptive' messages that show what you did wrong, and a pragma introducing a scope isnt really 'true' (so having it observably act as a scope seems wrong).

>These constructs, if enabled, should be pretty similar to lambdas, so I assume it should be fine to diagnose dangling break/continue here. The diagnostics might be improved, though.

I think the difference is that the lambda introduces a lookup scope/etc, whereas these pragmas aren't supposed to.  For example, setting the 'function' scope prevents something like:

```
auto foo(int i) {
#pragma acc parallel
int j = function_call();

return i + j;
}
```
from working. See : https://godbolt.org/z/Gq8Yen1vr (for the omp example).

This is actually something that is supposed to work in OpenACC, so introducing the scope makes that not work.

Reply:
Ok, I see. For OpenMP this is correct behavior per standard.

> And checking the 'is loop with a parent openacc scope' doesn't work if the openacc-pragma is inside the loop. 

Why it does not work?

Reply:
I was just thinking... since the concern here is readability: would it be acceptable to move the checking in this 'if' statement to a static function with multiple exits/a series of comments to describe what is happening?

Reply:
Sorry if I'm being dumb here, but I don't really get the question?  We're trying to avoid diagnosing when the `breakParent` is a `switch`, so we are excluding it.  If we were to NOT use the inversion, we would be diagnosing ONLY on a the switch (the opposite of what I want?).

So the logic for when to diagnose is:

`S->isOpenACCComputeConstructScope()` <-- Diagnose if the 'current' breakParent is a Compute Construct
OR
`(!S->isDirectlySwitchScope() && S->getParent() && S->getParent()->isOpenACCComputeConstructScope()))` <-- If the Parent of the `breakParent` is a Compute Construct, unless it is a `switch`.

Reply:
No problem we're in the same position. :)
I'm not asking about the very first check, it is ok and I di not ask about it.
What I'm asking is exactly the second part (after OR). I'm suggesting instead of checking for non-switch just to check explicitly for 'for scope'. "Not switch scope" is too broad, I just suggest to check for the scopes, which are not allowed instead (isForScope). I think it is much easier to read and understand the context.

Reply:
> No problem we're in the same position. :) I'm not asking about the very first check, it is ok and I did not ask about it. What I'm asking is exactly the second part (after OR, second part of the logic). I'm suggesting instead of checking for non-switch just to check explicitly for 'for scope'. "Not switch scope" is too broad, I just suggest to check for the scopes, which are not allowed instead (isForScope). I think it is much easier to read and understand the context.

Ah!  So we don't actually HAVE a `ForScope` as far as I can tell, let alone the `while` or `do-while`: https://clang.llvm.org/doxygen/classclang_1_1Scope.html

Perhaps I could change the name of `isDirectlySwitchScope` to be: `isBreakLoop` or something, and have it do the inverse of what `isDirectlySwitchScope` does now?

Reply:
> Perhaps I could change the name of isDirectlySwitchScope to be: isBreakLoop or something, and have it do the inverse of what isDirectlySwitchScope does now?

Yep, this will be much better!